install/upgrade.php

```diff
@@ -118,6 +118,12 @@
 $settings['bbname_orig'] = $settings['bbname'];
 $settings['bbname'] = strip_tags($settings['bbname']);
 
+// Adjust a relative upload path to an absolute one
+if(my_substr($settings['uploadspath'], 0, 1) != "/")
+{
+	$settings['uploadspath'] = MYBB_ROOT.$settings['uploadspath'];
+}
+
```

Comment:
As above, I suggest we replace this code with the code supplied above for better Windows compatibility.

Reply:
Why don't you just dump the controversy and go for `DIRECTORY_SEPARATOR`

Reply:
I don't understand. Can you explain in more detail?

Reply:
Its a constant of PHP which picks the right separator based on OS.

Reply:
Yep, yep, I know what it is, I just don't see how it helps here.

On Windows, an admin could use either a forward slash or a backslash when setting `uploadspath`. `DIRECTORY_SEPARATOR` isn't going to help us decide which was chosen by the admin, nor whether or not the admin chose to start their absolute path with a drive letter.

Reply:
Using `DIRECTORY_SEPARATOR` might not work perfectly either - path handling is hard 😢 

WordPress has a decent looking function that does what we're trying to do which might be worth having a glance at: https://developer.wordpress.org/reference/functions/path_is_absolute/

Reply:
We can run a preg_replace with `DIRECTORY_SEPARATOR` to whatever admin has set.
Can't we? Or I am missing something...

Reply:
> WordPress has a decent looking function that does what we're trying to do

Though it uses the same regex that I came up with independently, it looks flawed to me, in that the regex should only be tested on a Windows platform, since on a UNIX platform, C:\ is a valid name for a directory, and is not an absolute path.

Likewise for testing whether the first character is a backslash - that test should only be performed on Windows platform.

Would encourage you guys to look at the code I supplied and point out any errors that you see. It seems serviceable to me.

And I still don't see how `DIRECTORY_SEPARATOR` helps in any meaningful way. So, yes, I think you're missing something.

Reply:
:(
Attempt to remove the absolute part from the path (if its there it will be removed, if its not there it will be as is) and force prepend. 
I guess I have lost the track. If its about OS and separator its definitely not about absolute / relative then.
Side note: IIRC we do something with avatar upload, worth a look.

I will jump out of it now.

Reply:
> Would encourage you guys to look at the code I supplied and point out any errors that you see. It seems serviceable to me.

Will do, checking it in my email client wasn't ideal 😉

Reply:
> `if(!in_array(my_substr($settings['uploadspath'], 0, 1)), ["/","\"])`

The problems I see with that code are:

1. It doesn't recognise C:\ or D:\ or E:\ etc as the start of an absolute path on a Windows platform (which they are).
2. It mistakenly treats an opening backslash as indicating an absolute path on a UNIX platform.

Reply:
Thanks, @euantorano. The only improvement I can see is to recognise an opening file:// as indicating an absolute path - but are admins really going to enter a path to a file like that?

Reply:
Better detect the OS then and do comparison.

Reply:
> Better detect the OS then and do comparison.

...which is exactly what the code I supplied *does*. I'm guessing you missed it? Here it is again:

https://github.com/mybb/mybb/pull/4342#discussion_r607987647